src/dlr_relayvm.cc

```diff
@@ -48,10 +59,53 @@ void RelayVMModel::SetupVMModule() {
                     "to override TVM allocations. Using default allocators.";
   }
 
-  tvm::runtime::Module lib = tvm::runtime::Module::LoadFromFile(path_->model_lib);
-  std::ifstream relay_ob(path_->relay_executable, std::ios::binary);
-  std::string code_data((std::istreambuf_iterator<char>(relay_ob)),
-                        std::istreambuf_iterator<char>());
+  std::string code_data;
+  std::string model_lib_path;
+  std::string metadata_path;
+  std::string metadata_data;
+  for (DLRModelElem el : model_elems) {
+    if (el.type == DLRModelElemType::RELAY_EXEC) {
+      if (el.path != nullptr) {
+        std::ifstream relay_ob(el.path, std::ios::binary);
+        code_data.assign((std::istreambuf_iterator<char>(relay_ob)),
+                         std::istreambuf_iterator<char>());
+      } else if (el.data != nullptr && el.data_size > 0) {
+        code_data.assign(static_cast<const char*>(el.data), el.data_size);
+      } else {
+        throw dmlc::Error("Invalid RelayVM model element RELAY_EXEC");
+      }
+    } else if (el.type == DLRModelElemType::TVM_LIB) {
+      if (el.path != nullptr) {
+        model_lib_path = el.path;
+      } else {
+        throw dmlc::Error("Invalid RelayVM model element TVM_LIB");
+      }
+    } else if (el.type == DLRModelElemType::NEO_METADATA) {
+      if (el.path != nullptr) {
+        metadata_path = el.path;
+      } else if (el.data != nullptr) {
+        metadata_data = static_cast<const char*>(el.data);
+      } else {
+        throw dmlc::Error("Invalid model element NEO_METADATA");
+      }
+    }
+  }
+  if (code_data.empty() || model_lib_path.empty() ||
+      (metadata_path.empty() && metadata_data.empty())) {
+    throw dmlc::Error(
+        "Invalid RelayVM model. Must have RELAY_EXEC, TVM_LIB and NEO_METADATA elements");
+  }
+
+  // Load Metadata from file or String
+  if (!metadata_path.empty() && !IsFileEmpty(metadata_path)) {
+    LoadJsonFromFile(metadata_path, this->metadata_);
+    ValidateDeviceTypeIfExists();
+  } else if (!metadata_data.empty()) {
+    LoadJsonFromString(metadata_data, this->metadata_);
+    ValidateDeviceTypeIfExists();
```

Comment:
should we just do this up above? or just load the file from path into a string above, and always call LoadJsonFromString here?

Reply:
Removed `metadata_path`, added helper function `dlr::LoadFileToString`